syntax = "proto3";

package masterworker;


service MRWorker{
  rpc AssignMap(MapRequest) returns (MapReply) {}
  rpc AssignReduce (ReduceRequest) returns (ReduceReply) {}
}

message MapRequest {
  uint32 shard_id = 1;
  repeated File shard = 2;
  uint32 n_outputs = 3;
  string user_id = 4;
  string output_dir = 5;
}

message MapReply {
  bool succeed = 1;
  repeated string file_locs = 2;
}

message ReduceRequest {
  uint32 part_id = 1;
  repeated string file_locs = 2;
  string user_id = 3;
  uint32 n_outputs = 4;
  string output_dir = 5;
}

message ReduceReply {
  bool succeed = 1;
}

message File {
  string path = 1;
  int64 start_pos = 2;
  int64 end_pos = 3;
}
